### ISSUE TITLE
Unexpected properties added to array result

### ISSUE THREAD

**Issue by knolleary** [jsonata-js/jsonata#473]
When evaluating the expression `payload[true][]` against `{ "payload": ["one"] }` the resulting array has some unexpected properties added to it: ` [ 'one', sequence: true, keepSingleton: true ]`.

```
var jsonata = require("resonata");
var data = { payload: ["one"]}
var exp1 = jsonata("payload[true][]")

exp1.evaluate(data)
=> [ 'one', sequence: true, keepSingleton: true ]

```

This is the root cause of an issue raised against Node-RED: https://github.com/node-red/node-red/issues/2700 where the output of this expression was passed to another expression, and the result differed to when they passed what appeared to be the same payload straight to the second expression.

The Node-RED debug sidebar fails to display array properties, so they couldn't see the difference in the messages - that's a separate issue - but the expectation was the result would be the vanilla array without any extra properties on it.

Is that a reasonable expectation?

**Comment by andrew-coleman** [comment]
This is a duplicate of issue #296 
Adding these properties to arrays doesn't show up when the array is serialised as JSON, so in the context of JSONata, it is a convenient way of holding sequence metadata.  In much earlier releases, I had used hidden properties on the array but I found that this was having a significant performance impact (https://github.com/jsonata-js/jsonata/issues/296#issuecomment-453997682).  
I'm open to other ideas of how to fix this, as long as it doesn't impact performance.

**Comment by janvda** [comment]
@andrew-coleman ,  the thing is that these properties seem to have an unexpected impact when feeding into the jsonata expression `(payload)[]` that is used in the next node-red node (see [issue reported on node-red](https://github.com/node-red/node-red/issues/2700)).

So for me these properties are not the issues, but the fact that these properties are affecting jsonata expression further down the chain.

**Comment by knolleary** [comment]
@janvda that is an interesting point. Given JSONata's proper adherence to the JSON format, is it right that having properties on the Array (which is invalid JSON) should affect the result of the expression. Although that feels like a potential major breaking change if it were to stop doing that as I could well imagine user's relying on that behaviour today.

@andrew-coleman thanks for the link to the existing issue. I understand the performance concerns. Will take a look at what other approaches could be used. I'd rather not have to do sanitisation outside of JSONata as that would have to be run against every single result, whether it actually contains a modified array or not. I'm sure there must be a more elegant way of tracking that state (maybe a WeakMap type thing to associate metadata with objects). The question will be finding the performant fix.

**Comment by janvda** [comment]
@knolleary it is true that there might be node-red flows out there that count on the unexpected outcome of the jsonata expression.  But I think we all agree that the outcome produced by the jsonata expression (I am referring to the 2nd jsonata expression - not the first one which is outputting unexpected properties) is not the correct one.  So, I would classify this as a bug and not as a breaking change.  Note also that I don't think there are many flows out there that are exploiting this unexpected behaviour as I have used quite some jsonata and this is the first time I encountered this issue.

**Comment by knolleary** [comment]
@janvda please remember we're discussing this in the context of the JSONata library. JSONata is used much more widely then just Node-RED and what you do.

**Comment by janvda** [comment]
@knolleary I agree that we must discuss it here in the context of the JSONata library.  I sometimes forget that there is a world outside Node-RED :-) 

So focus back to ticket and its ticket description.

Ticket speaks about "unexpected properties added to array result" which indeed strongly looks like a duplicate of issue #296 

... but I am actually referring to a **more severe issue:** that is that some jsonata expressions no longer return correct results when getting such an array with unexpected properties as input.

I agree that this more severe issue is automatically solved when fixing the issue with unexpected properties, but there might be an easier/quicker/performant way to fix this more server problem without fixing the unexpected properties (#296 ).

As we have already ticket #296, I would like to limit the scope of this ticket to handle the more severe problem (= jsonata returning incorrect results when having as input the output of another jsonata expression that is containing those unexpected properties).

**Comment by andrew-coleman** [comment]
@janvda I'd be happy to, but can you give me an example of two expressions and the input data that exhibit this problem?

Preferably something I can recreate in Runkit (https://npm.runkit.com/jsonata)

E.g. something of the form:
```
var jsonata = require("jsonata")
var data = { payload: ["one"]}
var exp1 = jsonata("payload[true][]")  // first expression
var data1 = exp1.evaluate(data)
var exp2 = jsonata("$")    // second expression
var data2 = exp2.evaluate(data1)

console.log(data1, data2)
```
(although this works as expected)

**Comment by knolleary** [comment]
Here is a test case for the 'second' part of the issue.

```
var jsonata = require("jsonata");

var exp = jsonata("(payload)")

var data1 = { payload: ["one"]}

var data2 = { payload: ["one"]}
data2.payload.sequence = true;

var data3 = { payload: ["one"]}
data3.payload.foo = true;


console.log(exp.evaluate(data1));
// => [ "one" ]
console.log(exp.evaluate(data2));
// => "one"
console.log(exp.evaluate(data3));
// => [ 'one', foo: true ]

```

In summary, if JSONata is given an array with a `sequence` property already set (`data2` in the above example), then it treats it differently to one without that property set. This is specific to the property named `sequence` as it has special meaning in the internals of JSONata. (Note how `data3` is handled which has a property named `foo`).


I would suggest that if a fix for the original issue was found that removed the need to modify the array, then it would likely also mean it would no longer be looking for a `sequence` property and thus this issue would also get resolved with the same fix.

**Comment by janvda** [comment]
@andrew-coleman here below the case for which I reported https://github.com/node-red/node-red/issues/2700

```
var jsonata = require("jsonata")
var data = { payload: ["one"]}
var exp1 = jsonata("{ 'payload' : payload[true][]}")  // first expression
var data1 = exp1.evaluate(data)
var exp2 = jsonata("(payload)[]")    // second expression
var data2 = exp2.evaluate(data1)

var data3 = exp2.evaluate(data) 

console.log(data,data1,data2,data3)
```

![image](https://user-images.githubusercontent.com/17238398/93585687-740bb380-f9a7-11ea-8ecb-2b563db7f3fb.png)
